### PR TITLE
feat(projects): add projects endpoint

### PR DESCRIPTION
**Pull request by philloooo**
Description about what this pull request does.

Please make sure to follow the [DEV guidelines](https://gen3.org/resources/developer/dev-introduction/) before asking for review.

### New Features
- Implements /datasets/projects endpoint that has the same access control mechanism as /datasets

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
Jenkins Build 2 : time taken 42 min
Check the https://jenkins.planx-pla.net/job/perf-peregrine/job/PR-123/2/display/redirect


Test results:
	Passed: 109, Failed: 0, Skipped: 0



| Test | Time (`PR`) | Time (`master`) | Diff |
| - | - | - | - |
| @ExportPerformanceTests @Performance: Export all nodes of type: case | 0.443 | 0.264 | 0.18 |
| @ExportPerformanceTests @Performance: Export all nodes of type: experiment | 0.381 | 0.334 | 0.05 |
| @ExportPerformanceTests @Performance: Export all nodes of type: program | 0.068 | 0.052 | 0.02 |
| @ExportPerformanceTests @Performance: Export all nodes of type: project | 0.069 | 0.047 | 0.02 |
| @ExportPerformanceTests @Performance: Export all nodes of type: read_group_qc | 0.108 | 0.092 | 0.02 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: aliquot | 0.229 | 0.154 | 0.08 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: case | 0.314 | 0.288 | 0.03 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group | 0.092 | 0.077 | 0.02 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 0 | 0.691 | 0.655 | 0.04 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 2 | 0.55 | 0.523 | 0.03 |
| @QueryPerformanceTests @Performance: Executing topDown query # 1 | 0.086 | 0.076 | 0.01 |
| @QueryPerformanceTests @Performance: Executing topDown query # 4 | 0.316 | 0.271 | 0.05 |
| @QueryPerformanceTests @Performance: Executing topDown query # 5 | 0.394 | 0.353 | 0.04 |
| @SubmissionPerformanceTests @Performance: Submission aliquot 1000 | 28.237 | 22.073 | 6.16 |
| @SubmissionPerformanceTests @Performance: Submission clinical_test 100 | 2.568 | 2.4 | 0.17 |
| @SubmissionPerformanceTests @Performance: Submission diagnosis 100 | 3.185 | 2.353 | 0.83 |
| @SubmissionPerformanceTests @Performance: Submission family_history 100 | 2.697 | 2.106 | 0.59 |
| @SubmissionPerformanceTests @Performance: Submission publication 100 | 2.564 | 2.428 | 0.14 |
| @SubmissionPerformanceTests @Performance: Submission read_group_qc 10 | 0.719 | 0.626 | 0.09 |
| @SubmissionPerformanceTests @Performance: Submission sample 100 | 45.053 | 2.738 | 42.32 |
| @SubmissionPerformanceTests @Performance: Submission slide 100 | 33.849 | 2.358 | 31.49 |
| @SubmissionPerformanceTests @Performance: Submission slide_count 10 | 0.46 | 0.428 | 0.03 |
| @SubmissionPerformanceTests @Performance: Submission submitted_methylation 10 | 21.944 | 3.076 | 18.87 |
| @SubmissionPerformanceTests @Performance: Submission submitted_unaligned_reads 10 | 2.179 | 2.017 | 0.16 |
| @SubmissionPerformanceTests @Performance: Submission treatment 10 | 0.41 | 0.31 | 0.10 |

File: peregrine/blueprints/datasets.py [hunk 1]
```diff
@@ -60,3 +61,33 @@ def get_datasets():
...
+    else:
+        set_read_access_projects()
+    projects = flask.g.read_access_projects
+    if not projects:
+        return flask.jsonify({})
+    # construct a query that get counts for all projects
+    # because graphql can't add structure to group by projects,
+    # we labeled the count by project index and later parse it
+    # with regex to add structure to response
+    query = "{project { name code dbgap_accession_number "
```
Comment (zflamig) [inline peregrine/blueprints/datasets.py]: First: 0 here so this doesn't just report the first 10?

File: openapis/swagger.yaml [hunk 2]
```diff
@@ -42,6 +42,23 @@ paths:
...
+      tags:
+        - datasets
+      summary: Get high level information for all projects
+      responses:
+        '200':
+          description: high level information for all projects
+          content:
+            application/json:
+              schema:
+                type: object
```
Comment (zflamig) [inline openapis/swagger.yaml]:
```
type: array
    items:
      $ref: '#/components/schemas/Project'

components:
  schemas:
    Project:
      type: object
      properties:
        name:
          type: string
        code:
          type: string
        dbgap_accession_number:
          type: string
        description:
          type: string
        image_url:
          type: string
```

File: peregrine/blueprints/datasets.py [hunk 3]
```diff
@@ -60,3 +61,33 @@ def get_datasets():
...
+    for field in ['description', 'image_url']:
+        if dictionary.schema['project']['properties'].get(field):
+            query += field + ' '
+
+    query += '}}'
+    data, errors = graphql.execute_query(query, variables={})
+    if errors:
+        return flask.jsonify({"data": data, "errors": errors}), 400
+    return flask.jsonify({"projects": data["project"]})
```
Comment (zflamig) [inline peregrine/blueprints/datasets.py]: Lets just return `data["project"]` so this matches the swagger?

File: peregrine/blueprints/datasets.py [hunk 4]
```diff
@@ -60,3 +61,33 @@ def get_datasets():
...
+    to set PUBLIC_DATASETS to True, this endpoint will be open to
+    anonymous users
+    """
+    if os.environ.get("PUBLIC_DATASETS", "false").lower() == "true":
+        set_read_access_projects_for_public_endpoint()
+    else:
+        set_read_access_projects()
+    projects = flask.g.read_access_projects
+    if not projects:
+        return flask.jsonify({})
```
Comment (zflamig) [inline peregrine/blueprints/datasets.py]: Should be either empty array `[]` or `{projects: []}` depending on your answer to the comment below

File: peregrine/blueprints/datasets.py (hunk 3 above)
Comment (philloooo) [reply peregrine/blueprints/datasets.py]: it's more best practice to return object instead of array for json response. I will update in other places

**Comment by PlanXCyborg** [comment]
Jenkins Build 4 : time taken 37 min
Check the https://jenkins.planx-pla.net/job/perf-peregrine/job/PR-123/4/display/redirect


Test results:
	Passed: 109, Failed: 0, Skipped: 0



| Test | Time (`PR`) | Time (`master`) | Diff |
| - | - | - | - |
| @ExportPerformanceTests @Performance: Export all nodes of type: aliquot | 0.403 | 0.373 | 0.03 |
| @ExportPerformanceTests @Performance: Export all nodes of type: case | 0.397 | 0.255 | 0.14 |
| @ExportPerformanceTests @Performance: Export all nodes of type: experiment | 0.791 | 0.41 | 0.38 |
| @ExportPerformanceTests @Performance: Export all nodes of type: program | 0.181 | 0.053 | 0.13 |
| @ExportPerformanceTests @Performance: Export all nodes of type: project | 0.125 | 0.102 | 0.02 |
| @ExportPerformanceTests @Performance: Export all nodes of type: read_group | 0.778 | 0.536 | 0.24 |
| @ExportPerformanceTests @Performance: Export all nodes of type: read_group_qc | 0.172 | 0.113 | 0.06 |
| @ExportPerformanceTests @Performance: Export all nodes of type: sample | 0.584 | 0.531 | 0.05 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: aliquot | 0.28 | 0.175 | 0.11 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: case | 0.323 | 0.223 | 0.10 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: experiment | 0.289 | 0.203 | 0.09 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group | 0.169 | 0.07 | 0.10 |
| @ExportPerformanceTests @Performance: Exporting a record by ID on nodes of type: read_group_qc | 0.078 | 0.063 | 0.02 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 0 | 0.72 | 0.685 | 0.04 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 4 | 0.364 | 0.343 | 0.02 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 6 | 0.179 | 0.157 | 0.02 |
| @QueryPerformanceTests @Performance: Executing bottomUp query # 7 | 0.087 | 0.077 | 0.01 |
| @QueryPerformanceTests @Performance: Executing topDown query # 0 | 0.106 | 0.074 | 0.03 |
| @QueryPerformanceTests @Performance: Executing topDown query # 1 | 0.083 | 0.077 | 0.01 |
| @QueryPerformanceTests @Performance: Executing topDown query # 2 | 0.105 | 0.095 | 0.01 |
| @QueryPerformanceTests @Performance: Executing topDown query # 4 | 0.373 | 0.249 | 0.12 |
| @QueryPerformanceTests @Performance: Executing topDown query # 5 | 0.401 | 0.332 | 0.07 |
| @QueryPerformanceTests @Performance: Executing topDown query # 6 | 0.528 | 0.439 | 0.09 |
| @QueryPerformanceTests @Performance: Executing topDown query # 7 | 0.562 | 0.514 | 0.05 |
| @SubmissionPerformanceTests @Performance: Submission acknowledgement 10 | 0.502 | 0.332 | 0.17 |
| @SubmissionPerformanceTests @Performance: Submission acknowledgement 1000 | 26.049 | 22.248 | 3.80 |
| @SubmissionPerformanceTests @Performance: Submission aligned_reads_index 10 | 2.256 | 1.999 | 0.26 |
| @SubmissionPerformanceTests @Performance: Submission aligned_reads_index 100 | 21.586 | 15.838 | 5.75 |
| @SubmissionPerformanceTests @Performance: Submission aliquot 100 | 2.73 | 2.257 | 0.47 |
| @SubmissionPerformanceTests @Performance: Submission clinical_test 100 | 2.846 | 2.543 | 0.30 |
| @SubmissionPerformanceTests @Performance: Submission core_metadata_collection 10 | 1.097 | 0.887 | 0.21 |
| @SubmissionPerformanceTests @Performance: Submission diagnosis 10 | 0.496 | 0.381 | 0.11 |
| @SubmissionPerformanceTests @Performance: Submission diagnosis 100 | 3.177 | 2.683 | 0.49 |
| @SubmissionPerformanceTests @Performance: Submission experimental_metadata 10 | 2.336 | 1.661 | 0.67 |
| @SubmissionPerformanceTests @Performance: Submission experimental_metadata 100 | 20.196 | 15.793 | 4.40 |
| @SubmissionPerformanceTests @Performance: Submission exposure 10 | 0.347 | 0.328 | 0.02 |
| @SubmissionPerformanceTests @Performance: Submission exposure 1000 | 27.728 | 23.998 | 3.73 |
| @SubmissionPerformanceTests @Performance: Submission family_history 10 | 0.618 | 0.323 | 0.29 |
| @SubmissionPerformanceTests @Performance: Submission family_history 100 | 2.671 | 2.335 | 0.34 |
| @SubmissionPerformanceTests @Performance: Submission keyword 10 | 0.455 | 0.37 | 0.09 |
| @SubmissionPerformanceTests @Performance: Submission keyword 100 | 2.553 | 2.205 | 0.35 |
| @SubmissionPerformanceTests @Performance: Submission publication 100 | 3.227 | 2.237 | 0.99 |
| @SubmissionPerformanceTests @Performance: Submission read_group_qc 10 | 0.676 | 0.49 | 0.19 |
| @SubmissionPerformanceTests @Performance: Submission slide 100 | 2.73 | 2.533 | 0.20 |
| @SubmissionPerformanceTests @Performance: Submission slide_image 10 | 2.169 | 1.716 | 0.45 |
| @SubmissionPerformanceTests @Performance: Submission slide_image 100 | 20.495 | 15.556 | 4.94 |
| @SubmissionPerformanceTests @Performance: Submission submitted_aligned_reads 10 | 2.09 | 1.654 | 0.44 |
| @SubmissionPerformanceTests @Performance: Submission submitted_aligned_reads 100 | 23.699 | 15.881 | 7.82 |
| @SubmissionPerformanceTests @Performance: Submission submitted_copy_number 10 | 2.282 | 1.635 | 0.65 |
| @SubmissionPerformanceTests @Performance: Submission submitted_copy_number 100 | 19.133 | 16.57 | 2.56 |
| @SubmissionPerformanceTests @Performance: Submission submitted_methylation 100 | 22.475 | 15.884 | 6.59 |
| @SubmissionPerformanceTests @Performance: Submission submitted_somatic_mutation 10 | 2.246 | 1.645 | 0.60 |
| @SubmissionPerformanceTests @Performance: Submission submitted_somatic_mutation 100 | 23.275 | 15.795 | 7.48 |
| @SubmissionPerformanceTests @Performance: Submission submitted_unaligned_reads 10 | 2.046 | 1.717 | 0.33 |
| @SubmissionPerformanceTests @Performance: Submission submitted_unaligned_reads 100 | 19.348 | 15.845 | 3.50 |

**Pull request merged**